File: windows/MotionMouse/Protocol/PacketParser.cs
using System.IO;
using System.Buffers.Binary;
using System.Text;
using Microsoft.Extensions.Logging;

namespace MotionMouse.Protocol;

/// <summary>
/// Parsed representation of an incoming packet from the Android app.
/// </summary>
public abstract record IncomingPacket
{
    /// <summary>
    /// HELLO (0x01) — Android is initiating a connection.
    /// </summary>
    public record Hello(int ProtocolVersion, string DeviceName) : IncomingPacket;

    /// <summary>
    /// MOTION (0x03) — Cursor velocity update.
    /// DeltaX and DeltaY are in pixels per second.
    /// TimestampMs is the Android send time for latency calculation.
    /// </summary>
    public record Motion(float DeltaX, float DeltaY, long TimestampMs) : IncomingPacket;

    /// <summary>
    /// BUTTON (0x04) — Mouse button press or release.
    /// </summary>
    public record Button(byte ButtonId, byte Action) : IncomingPacket;

    /// <summary>
    /// PING (0x05) — Android requesting latency measurement.
    /// </summary>
    public record Ping(long SendTimestampMs) : IncomingPacket;

    /// <summary>
    /// STATUS (0x08) — Battery level, lock state, connection type.
    /// </summary>
    public record Status(
        int BatteryLevel,
        bool IsLocked,
        byte ConnectionType
    ) : IncomingPacket;

    /// <summary>
    /// DISCONNECT (0x09) — Android is disconnecting cleanly.
    /// </summary>
    public record Disconnect : IncomingPacket;

    /// <summary>
    /// An unrecognised packet type — log and ignore.
    /// Forward compatibility for future Android versions.
    /// </summary>
    public record Unknown(byte TypeId) : IncomingPacket;
}

/// <summary>
/// Parses incoming binary packets from the Android application.
///
/// Reads from a Stream using length-prefix framing identical to
/// the Android PacketParser — see that file for framing rationale.
///
/// All methods are async to avoid blocking the UI thread.
/// The caller (transport receive loop) runs on a background thread,
/// but async keeps the pattern consistent and enables future
/// cancellation support.
/// </summary>
public sealed class PacketParser
{
    private readonly ILogger<PacketParser> _logger;

    // Protocol framing constants
    private const int LengthFieldSize = 2;
    private const int MaxPacketSize   = 64;

    public PacketParser(ILogger<PacketParser> logger)
    {
        _logger = logger;
    }

    /// <summary>
    /// Read and parse the next complete packet from the stream.
    ///
    /// Blocks asynchronously until data is available.
    /// Returns null when the stream is closed cleanly.
    /// Throws IOException on unexpected connection loss.
    /// </summary>
    public async Task<IncomingPacket?> ReadNextPacketAsync(
        Stream stream,
        CancellationToken cancellationToken = default)
    {
        // Step 1: Read 2-byte length prefix
        var lengthBuffer = new byte[LengthFieldSize];
        if (!await ReadFullyAsync(stream, lengthBuffer, cancellationToken))
            return null; // Clean close

        // Big-endian uint16 — matches Android's ByteBuffer.BIG_ENDIAN
        var totalLength = BinaryPrimitives.ReadUInt16BigEndian(lengthBuffer);

        if (totalLength < 3 || totalLength > MaxPacketSize)
        {
            _logger.LogWarning("Invalid packet length {Length} — dropping connection", totalLength);
            return null;
        }

        // Step 2: Read remaining bytes
        var remaining = new byte[totalLength - LengthFieldSize];
        if (!await ReadFullyAsync(stream, remaining, cancellationToken))
            return null;

        // Step 3: Parse type and payload
        var typeId = remaining[0];
        _logger.LogDebug("Received packet: Type=0x{TypeId:X2}, TotalLength={TotalLength}", typeId, totalLength);
        var payload = remaining.AsSpan(1); // Everything after the type byte

        return typeId switch
        {
            PacketType.Hello        => ParseHello(payload),
            PacketType.Motion       => ParseMotion(payload),
            PacketType.Button       => ParseButton(payload),
            PacketType.Ping         => ParsePing(payload),
            PacketType.Status       => ParseStatus(payload),
            PacketType.Disconnect   => new IncomingPacket.Disconnect(),
            _ => new IncomingPacket.Unknown(typeId)
        };
    }

    // --- Individual packet parsers ---

    private IncomingPacket ParseHello(ReadOnlySpan<byte> payload)
    {
        try
        {
            var version = payload[0];
            var nameLen = payload[1];

            if (version == 0xFF)
            {
                _logger.LogError("Protocol version mismatch from Android client");
                return new IncomingPacket.Unknown(PacketType.Hello);
            }

            var name = Encoding.UTF8.GetString(payload.Slice(2, nameLen));
            return new IncomingPacket.Hello(version, name);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to parse HELLO packet");
            return new IncomingPacket.Unknown(PacketType.Hello);
        }
    }

    private IncomingPacket ParseMotion(ReadOnlySpan<byte> payload)
    {
        // Payload: [4] float deltaX, [4] float deltaY, [8] long timestampMs
        // All big-endian per protocol spec
        var deltaX      = BinaryPrimitives.ReadSingleBigEndian(payload[..4]);
        var deltaY      = BinaryPrimitives.ReadSingleBigEndian(payload[4..8]);
        var timestampMs = BinaryPrimitives.ReadInt64BigEndian(payload[8..16]);
        return new IncomingPacket.Motion(deltaX, deltaY, timestampMs);
    }

    private IncomingPacket ParseButton(ReadOnlySpan<byte> payload)
    {
        return new IncomingPacket.Button(payload[0], payload[1]);
    }

    private IncomingPacket ParsePing(ReadOnlySpan<byte> payload)
    {
        var timestamp = BinaryPrimitives.ReadInt64BigEndian(payload[..8]);
        return new IncomingPacket.Ping(timestamp);
    }

    private IncomingPacket ParseStatus(ReadOnlySpan<byte> payload)
    {
        return new IncomingPacket.Status(
            BatteryLevel:   payload[0],
            IsLocked:       payload[1] == 0x01,
            ConnectionType: payload[2]
        );
    }

    // --- Stream reading utility ---

    /// <summary>
    /// Read exactly buffer.Length bytes from stream into buffer.
    /// Returns false on clean stream end (EOF).
    /// Throws IOException on unexpected error.
    /// </summary>
    private static async Task<bool> ReadFullyAsync(
        Stream stream,
        byte[] buffer,
        CancellationToken cancellationToken)
    {
        var offset = 0;
        while (offset < buffer.Length)
        {
            var read = await stream.ReadAsync(
                buffer.AsMemory(offset, buffer.Length - offset),
                cancellationToken);

            if (read == 0)
                return false; // Clean EOF

            offset += read;
        }
        return true;
    }
}

File: android/app/src/main/kotlin/com/motionmouse/app/protocol/PacketParser.kt
package com.motionmouse.app.protocol

import android.util.Log
import java.io.EOFException
import java.io.InputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder

private const val TAG = "PacketParser"

/**
 * Parses incoming binary packets from the Windows application.
 *
 * Reads from a blocking InputStream (TCP socket stream).
 * Must be called from a dedicated IO thread — it blocks on read().
 *
 * Framing strategy:
 *   1. Read exactly 2 bytes → total packet length
 *   2. Read exactly (length - 2) remaining bytes
 *   3. Parse type byte and route to appropriate handler
 *   4. Return sealed IncomingPacket
 *
 * This length-prefix framing ensures we always read exactly one
 * complete packet regardless of how TCP segments the stream.
 * TCP is a stream protocol — without framing, a single read()
 * might return half a packet or two packets concatenated.
 *
 * Error handling:
 *   - EOFException → connection closed cleanly → return null
 *   - IOException  → connection lost → propagate to caller
 *   - Malformed packet → log and return Unknown (don't crash)
 *   - Unknown type → return Unknown (forward compatibility)
 */
object PacketParser {

    private const val LENGTH_FIELD_SIZE = 2
    private const val MAX_PACKET_SIZE = 64  // Protocol enforced maximum

    /**
     * Read and parse the next packet from the stream.
     *
     * Blocks until a complete packet is available.
     *
     * @return Parsed IncomingPacket, or null if the connection was
     *         closed cleanly (EOF).
     * @throws java.io.IOException if the connection was lost unexpectedly.
     */
    fun readNextPacket(inputStream: InputStream): IncomingPacket? {
        // Step 1: Read 2-byte length prefix
        val lengthBytes = ByteArray(LENGTH_FIELD_SIZE)
        try {
            readFully(inputStream, lengthBytes)
        } catch (e: EOFException) {
            // Clean connection close — not an error
            return null
        }

        val totalLength = ByteBuffer.wrap(lengthBytes)
            .order(ByteOrder.BIG_ENDIAN)
            .short
            .toInt() and 0xFFFF  // Treat as unsigned

        // Validate length before allocating
        if (totalLength < LENGTH_FIELD_SIZE + 1 || totalLength > MAX_PACKET_SIZE) {
            Log.w(TAG, "Invalid packet length: $totalLength — dropping connection")
            return null  // Caller should close socket
        }

        // Step 2: Read remaining bytes (everything after the length prefix)
        val remainingSize = totalLength - LENGTH_FIELD_SIZE
        val remainingBytes = ByteArray(remainingSize)
        readFully(inputStream, remainingBytes)

        // Step 3: Parse type and payload
        val buffer = ByteBuffer.wrap(remainingBytes).order(ByteOrder.BIG_ENDIAN)
        val typeId = buffer.get()

        return when (typeId) {
            PacketType.HELLO_ACK    -> parseHelloAck(buffer)
            PacketType.SETTINGS_SYNC -> parseSettingsSync(buffer)
            PacketType.PONG         -> parsePong(buffer)
            PacketType.DISCONNECT   -> IncomingPacket.Disconnect
            else -> {
                Log.d(TAG, "Unknown packet type: 0x${typeId.toString(16)} — ignoring")
                IncomingPacket.Unknown(typeId)
            }
        }
    }

    // --- Individual packet parsers ---

    /**
     * HELLO_ACK (0x02)
     * Payload: [1] version, [1] nameLen, [N] pcName
     */
    private fun parseHelloAck(buffer: ByteBuffer): IncomingPacket {
        return try {
            val version = buffer.get().toInt() and 0xFF
            val nameLen = buffer.get().toInt() and 0xFF
            val nameBytes = ByteArray(nameLen)
            buffer.get(nameBytes)
            val pcName = String(nameBytes, Charsets.UTF_8)

            if (version == 0xFF) {
                // Windows rejected our protocol version
                Log.e(TAG, "Protocol version mismatch — server rejected connection")
                IncomingPacket.Unknown(PacketType.HELLO_ACK)
            } else {
                IncomingPacket.HelloAck(version, pcName)
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse HELLO_ACK", e)
            IncomingPacket.Unknown(PacketType.HELLO_ACK)
        }
    }

    /**
     * SETTINGS_SYNC
     * Payload: [4] sensitivityX, [4] sensitivityY, [4] smoothing, [4] deadZone, [4] accelExponent
     */
    private fun parseSettingsSync(buffer: ByteBuffer): IncomingPacket {
        return try {
            IncomingPacket.SettingsSync(
                sensitivityX = buffer.float,
                sensitivityY = buffer.float,
                smoothingFactor = buffer.float,
                deadZone = buffer.float,
                accelerationExponent = buffer.float
            )
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse SETTINGS_SYNC", e)
            IncomingPacket.Unknown(PacketType.SETTINGS_SYNC)
        }
    }

    /**
     * PONG (0x06)
     * Payload: [8] original send timestamp
     */
    private fun parsePong(buffer: ByteBuffer): IncomingPacket {
        return try {
            IncomingPacket.Pong(originalSendTimestamp = buffer.long)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse PONG", e)
            IncomingPacket.Unknown(PacketType.PONG)
        }
    }

    // --- Stream reading utility ---

    /**
     * Read exactly [length] bytes from [stream] into [buffer].
     *
     * InputStream.read() is not guaranteed to return all requested bytes
     * in a single call — it may return fewer. This is especially common
     * with TCP sockets. readFully() loops until we have exactly what we need.
     *
     * @throws EOFException if the stream ends before we read enough bytes.
     * @throws IOException  if the stream throws during read.
     */
    private fun readFully(stream: InputStream, buffer: ByteArray) {
        var offset = 0
        while (offset < buffer.size) {
            val bytesRead = stream.read(buffer, offset, buffer.size - offset)
            if (bytesRead == -1) throw EOFException("Stream ended after $offset bytes")
            offset += bytesRead
        }
    }
}
